src/bulk_typer/actions.clj

```diff
@@ -0,0 +1,67 @@
+(ns bulk-typer.actions
+  (:require [clojure.math.numeric-tower :as math]
+            [clojure-commons.file-utils :as ft]
+            [clj-jargon.init :as init]
+            [clj-icat-direct.icat :as icat]
+            [clojure.string :as string]
+            [bulk-typer.irods :as irods]
+            [bulk-typer.config :as cfg])
+  (:import [java.util.concurrent Executors]))
+
+(def irods-pool (Executors/newFixedThreadPool 5))
+(def icat-pool (Executors/newFixedThreadPool 5))
+
+(defn- mk-jargon-cfg
+  []
+  (init/init (cfg/irods-host)
+             (cfg/irods-port)
+             (cfg/irods-user)
+             (cfg/irods-pass)
+             (cfg/irods-home)
+             (cfg/irods-zone)
+             (cfg/irods-resc)
+             :max-retries (cfg/irods-max-retries)
+             :retry-sleep (cfg/irods-retry-sleep)
+             :use-trash   (cfg/irods-use-trash)))
+
+(defn- make-prefixes
+  ([]
+   (make-prefixes (cfg/base-prefix-length)))
+  ([length]
+   (let [format-str (format "%%0%dx" length)]
+   (map (fn [x] (format format-str x)) (shuffle (range 0 (math/expt 16 length)))))))
+
+(defn- do-files
+  [files]
+  (init/with-jargon (mk-jargon-cfg) [cm]
+    (let [;; create an agent, queue loading data, and queue getting the file type from that data
+          agents (mapv (fn [f]
+                         (as-> (agent f) a
+                               (send-via irods-pool a (fn [f] [f (irods/get-data cm f)]))
+                               (send a (fn [[f d]] [f (irods/get-file-type d f)]))
+                               (send-via icat-pool a (fn [[f t]] [f t (irods/add-type-if-unset cm f t)])))) files)]
```

Comment:
Agents are super cool -- this gets the data in one (5-thread) pool, detects the type in a pool based on available CPUs, then adds the type in another 5-thread pool. The `await` below makes sure these are all done. The limit of 5 means that we won't have to suffer too many extra connections overhead, since jargon is weird about threads.

Reply:
Nice 🎉 ✨